src/ip_address.h

```diff
@@ -0,0 +1,158 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+ * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ */
+
+#ifndef IP_ADDRESS_H_
+#define IP_ADDRESS_H_
+
+#include <sys/socket.h>
+#include <netinet/in.h>
+
+#include <functional>
+#include <string>
+
+#ifndef __FreeBSD__
+#include <tr1/unordered_map>
+#else
+#include <unordered_map>
+#endif
+
+class IPAddress {
+private:
+    sa_family_t m_family;
+    union {
+        in_addr m_addr4;
+        in6_addr m_addr6;
+    };
+
+public:
+    static bool resolve(const char *str, IPAddress &out, std::string &err);
+    static IPAddress zero();
+
+    IPAddress() : m_family(AF_UNSPEC)
+    {
+    }
+
+    IPAddress(const IPAddress &rhs) : m_family(rhs.m_family), m_addr6(rhs.m_addr6)
+    {
+    }
+
+    IPAddress(const sockaddr *sa, socklen_t len)
+    {
+        if (len < sizeof(m_family)) {
+            m_family = AF_UNSPEC;
+            return;
+        }
+        m_family = sa->sa_family;
+        switch (m_family) {
+        case AF_INET:
+            if (len >= sizeof(sockaddr_in)) {
+                m_addr4 = ((sockaddr_in *)sa)->sin_addr;
+            } else {
+                m_family = AF_UNSPEC;
+            }
+            break;
+        case AF_INET6:
+            if (len >= sizeof(sockaddr_in6)) {
+                m_addr6 = ((sockaddr_in6 *)sa)->sin6_addr;
+            } else {
+                m_family = AF_UNSPEC;
+            }
+            break;
+        }
+    }
+
+    inline sa_family_t family() const
+    {
+        return m_family;
+    }
+
+    inline const in_addr &addr4() const
+    {
+        return m_addr4;
+    }
+
+    inline const in6_addr &addr6() const
+    {
+        return m_addr6;
+    }
+
+    bool is_specified() const;
+
+    std::string toString() const;
+
+    inline friend bool operator==(const IPAddress &key1, const IPAddress &key2)
+    {
+        if (key1.m_family == key2.m_family) {
+            switch (key1.m_family) {
+            case AF_UNSPEC:
+                return true;
+            case AF_INET:
+                return key1.m_addr4.s_addr == key2.m_addr4.s_addr;
+            case AF_INET6:
+                return IN6_ARE_ADDR_EQUAL(&key1.m_addr6, &key2.m_addr6);
+            }
+        }
+        return false;
+    }
+};
+
+// The only types that TR1 has built-in hash/equal_to functions for, are scalar types,
+// std:: string, and std::wstring. For any other type, we need to write a
+// hash/equal_to functions, by ourself.
+namespace std {
+#if !defined(WIN32) && !defined(__FreeBSD__)
+namespace tr1 {
+#endif
+template <> struct hash<IPAddress> : public std::unary_function<IPAddress, int> {
+    int operator()(IPAddress const &key) const
+    {
+        switch (key.family()) {
+        case AF_INET:
+            return key.addr4().s_addr & 0xFF;
```

Comment:
Why do we use only 1 byte?

Reply:
We are using this hash function for years. I just moved it into another place of code.